### PR TITLE
consensus: document merkle mutation root invariant

### LINKED ISSUE #28457: consensus: Better document ComputeMerkleRoot & add test for return value
At least twice we've had attempts to refactor [`ComputeMerkleRoot`](https://github.com/bitcoin/bitcoin/blob/fd69ffbbfb3e08b474b33540e56cf4f81e5c21d4/src/consensus/merkle.cpp#L45) (#22046 & #22046). While the value of any sort of refactor is dubious, it may also come with subtle, unintended side-effects, [see review comment here](https://github.com/bitcoin/bitcoin/pull/28430#issuecomment-1714504535):

> 1) The added `break`  only saves time if we receive a mutated block with repeated transactions. So this change isn't relevant unless a peer send us a block they mutated on purpose, in which case they get disconnected for that. So it really shouldn't change anything in a  non-adversarial scenario.

> 2) If we break early and don't compute the full merkle root anymore, the return value of `BlockMerkleRoot()` changes. As a result, [code calling BlockMerkleRoot()](https://github.com/bitcoin/bitcoin/blob/8f7b9eb8711fdb32e8bdb59d2a7462a46c7a8086/src/validation.cpp#L3534-L3542) https://github.com/bitcoin/bitcoin/blob/8f7b9eb8711fdb32e8bdb59d2a7462a46c7a8086/src/validation.cpp#L3534-L3542  would now return "bad-txnmrklroot" as an error instead of "bad-txns-duplicate", because now the first check already fails. This seems incorrect.

It might be good to better document this code, and add a test that would fail if a similar refactor is attempted in future.

### PR DESCRIPTION
**Pull request by l0rinc**
**Problem:** `ComputeMerkleRoot`'s optional mutation flag and the reasoning behind its per-level check are undocumented, and the behavior is only exercised indirectly by merkle_test through random duplications and old-vs-new comparisons, so a refactor could silently change it, as the discussions in #22046 and #28430 illustrate.

**Fix:** Document the flag on the function declaration, explain inside the inner loop why the mutation check runs at every tree level even after a duplicate is found, and add direct API coverage for the CVE-2012-2459 construction.

**Coverage check:** Both `merkle_test` and the new `merkle_test_mutated_return_value` would fail under a refactor that stops the outer reduction once mutation is detected, e.g.:
<details><summary>Hypothetical regression</summary>

```patch
diff --git a/src/consensus/merkle.cpp b/src/consensus/merkle.cpp
index dfa23cf897..40bc3f8efa 100644
--- a/src/consensus/merkle.cpp
+++ b/src/consensus/merkle.cpp
@@ -59,6 +59,7 @@ uint256 ComputeMerkleRoot(std::vector<uint256> hashes, bool* mutated) {
                 if (hashes[pos] == hashes[pos + 1]) mutation = true;
             }
         }
+        if (mutation) break;
         if (hashes.size() & 1) {
             hashes.push_back(hashes.back());
         }
```
</details>

Fixes #28457

### CODE REVIEW

**Review by optout21 — COMMENTED** [review]
Concept ACK f2dbc6a5fd582bb125c2a277362a7fe53ea7aa17

Comments added to the `ComputeMerkleRoot` method, and a new test case with one corner-case test added.

**Comment by w0xlt** [comment]
Concept ACK

**Review by optout21 — COMMENTED** [review]
ACK f2dbc6a5fd582bb125c2a277362a7fe53ea7aa17

The change documents the method `ComputeMerkleRoot`, including a CVE-relevant corner case. No behavior change, only comments and test code is affected.

Since `ComputeMerkleRoot` is being documented here, I would suggest to also include:
- that for optimal performance `hashes` should be reserved to an even size (#32497).
- the two discarded historical minor optimization opportunities (#22046 & #28430; PR numbers without details).

**Comment by l0rinc** [comment]
> that for optimal performance hashes should be reserved to an even size (https://github.com/bitcoin/bitcoin/pull/32497).

That's just an optimization, it would work without the preallocation as well

> the two discarded historical minor optimization opportunities (https://github.com/bitcoin/bitcoin/pull/22046 & https://github.com/bitcoin/bitcoin/pull/28430; PR numbers without details).

Not sure, these docs were added to discourage tinkering with that part without a very good reason.

**Comment by sedited** [comment]
@w0xlt do you want to take another look here?

File: src/consensus/merkle.cpp
Comment (hodlinator) [inline src/consensus/merkle.cpp:52]: nanonit: The above comment arguably applies to the entire file, but I think it would be preferable to have it adjacent to the core function implementing what it describes.
```suggestion
*/
uint256 ComputeMerkleRoot(std::vector<uint256> hashes, bool* mutated) {
```

File: src/test/merkle_tests.cpp [hunk 1]
```diff
@@ -200,6 +200,20 @@ BOOST_AUTO_TEST_CASE(merkle_test_OddTxWithRepeatedLastTx_block)
...
 
+BOOST_AUTO_TEST_CASE(merkle_test_mutated_return_value)
+{
+    // CVE-2012-2459 construction: [1,2,3,4,5,6] and [1,2,3,4,5,6,5,6] produce the same root.
+    const std::vector leaves{uint256{1}, uint256{2}, uint256{3}, uint256{4}, uint256{5}, uint256{6}};
+    auto mutated_leaves{leaves};
+    mutated_leaves.insert(mutated_leaves.end(), leaves.end() - 2, leaves.end()); // repeat last two elements
+
+    bool mutated{true};
+    const auto unmutated_root{ComputeMerkleRoot(leaves, &mutated)};
```
Comment (hodlinator) [inline src/test/merkle_tests.cpp]: nanonit: Preference for calling out the type here:
```suggestion
    const uint256 unmutated_root{ComputeMerkleRoot(leaves, &mutated)};
```

File: src/consensus/merkle.cpp [hunk 2]
```diff
@@ -47,6 +47,8 @@ uint256 ComputeMerkleRoot(std::vector<uint256> hashes, bool* mutated) {
     bool mutation = false;
     while (hashes.size() > 1) {
         if (mutated) {
+            // Mutation detection only affects the output flag.
+            // The root computation below must still reduce every tree level.
```
Comment (hodlinator) [inline src/consensus/merkle.cpp]: Something like the following seems more helpful to me:
```suggestion
            // The mutation check needs to happen at every level, not just the
            // lowest one, as that would not catch for example [1,2,1,2].
            // We repeat the check without first checking if mutation has
            // already been detected since blocks with mutated transaction
            // lists should not propagate through the network anyway, and
            // blocks that are invalid in this way are not more expensive to
            // process than valid ones.
```

**Review by hodlinator — COMMENTED** [review]
Concept ACK f2dbc6a5fd582bb125c2a277362a7fe53ea7aa17

File: src/consensus/merkle.cpp (hunk 2 above)
Comment (l0rinc) [reply src/consensus/merkle.cpp]: Thanks, added something similar

Comment (hodlinator) [reply src/consensus/merkle.cpp]: Thanks!

Don't you think that mutated blocks not propagating is also a factor in why it's not worth optimizing the mutation check by not repeating it if already detected? See point 1 in https://github.com/bitcoin/bitcoin/pull/28430#issuecomment-1714504535.

File: src/test/merkle_tests.cpp [hunk 3]
```diff
@@ -200,6 +200,26 @@ BOOST_AUTO_TEST_CASE(merkle_test_OddTxWithRepeatedLastTx_block)
...
+BOOST_AUTO_TEST_CASE(merkle_test_mutated_return_value)
+{
+    // CVE-2012-2459 construction: [1,2,3,4,5,6] and [1,2,3,4,5,6,5,6] produce the same root.
+    const std::vector leaves{uint256{1}, uint256{2}, uint256{3}, uint256{4}, uint256{5}, uint256{6}};
+    auto mutated_leaves{leaves};
+    mutated_leaves.insert(mutated_leaves.end(), leaves.end() - 2, leaves.end()); // repeat last two elements
+
+    bool mutated{true};
+    const uint256 unmutated_root{ComputeMerkleRoot(leaves, &mutated)};
+    BOOST_CHECK_EQUAL(mutated, false);
```
Comment (hodlinator) [inline src/test/merkle_tests.cpp]: Thanks for revealing the types!

nit: Comparing things to boolean literals to see if the boolean result value is true is too verbose in my book.
```
BOOST_CHECK_EQUAL(mutated, false):
../merkle_tests.cpp(220): error: in "merkle_tests/...": check mutated == false has failed [true != false]

BOOST_CHECK(!mutated):
../merkle_tests.cpp(220): error: in "merkle_tests/...": check !mutated has failed

[On parent master commit 625f951ba2fcd0b7c66de8bb30244b0b2fa63673]

₿ git grep -E "BOOST_CHECK_EQUAL\(\w+, (false|true)\);" | wc -l
59

₿ git grep -E "BOOST_CHECK\(" | wc -l
4087
```

File: src/consensus/merkle.cpp
Comment (hodlinator) [inline src/consensus/merkle.cpp:1]: The commit message and PR description are quite verbose. Point 2 quoted in #28457 about getting a different merkle root value was stricken out in the original comment https://github.com/bitcoin/bitcoin/pull/28430#issuecomment-1714504535 as it was based on a *misunderstanding of the effects* of the suggested change. So I don't think it's really something that needs to be called out. It's pretty axiomatic that a function should return correctly computed values, the added unit test should suffice. The main part I suggest should be removed/shortened is:

> `ComputeMerkleRoot` reports duplicate-subtree mutations through an optional output flag that is always written when provided, but callers still rely on the return value being the Merkle root for the whole input.

I also think this comment in the .CPP is redundant:

```c++
// Mutation detection only affects the output flag; root computation
// must still reduce the full input.
```

And this one in the .H:

```c++
 * ... The returned root is always computed from the full input,
 * regardless of whether a mutation is detected.
```

It's worth documenting *surprising* behavior, but starting to bloat the codebase with too many redundant comments harms readability IMO. Especially when it's based on a retracted misunderstanding of a closed PR.

**Review by hodlinator — COMMENTED** [review]
Reviewed 86e115927a287dc326aa3a167f4b889191692197

File: src/test/merkle_tests.cpp (hunk 3 above)
Comment (l0rinc) [reply src/test/merkle_tests.cpp]: I usually do that when I have we have both `true` and `false` cases nearby (and the `!` may not immediately be obvious, but that can also be clarified with a leading space).
Or when I'm doing a characterization test so that the fix commit just changes `true` to `false` instead of adding a `!` - but here it's not the case, so I'll revert, thanks.

**Review by w0xlt — COMMENTED** [review]
ACK 86e115927a287dc326aa3a167f4b889191692197
The documentation matches the existing mutation and root-computation behavior.

File: src/consensus/merkle.cpp
Comment (l0rinc) [reply src/consensus/merkle.cpp:1]: Thanks, pushed

File: src/consensus/merkle.cpp (hunk 2 above)
Comment (l0rinc) [reply src/consensus/merkle.cpp]: Added, thanks

**Review by w0xlt — COMMENTED** [review]
reACK ef501a63d9d65fce49bd633c424e0bbabead7ef6

**Review by hodlinator — APPROVED** [review]
ACK ef501a63d9d65fce49bd633c424e0bbabead7ef6

* Adds unit test to enforce return value & out parameters have expected behavior.
* Adds comments explaining why the code is written that way, which also indirectly explains why the optimization suggestions in #22046 and #28430 were rejected.

---

Thanks for incorporating my feedback!

**Comment by optout21** [comment]
reACK ef501a63d9d65fce49bd633c424e0bbabead7ef6

**Comment by achow101** [comment]
ACK ef501a63d9d65fce49bd633c424e0bbabead7ef6
